pkg/ccl/backupccl/backup_test.go

```diff
@@ -1443,6 +1443,252 @@ into_db='restoredb', %s)`, encryptionOption), backupLoc1)
+				sqlDB.Exec(t, `SET CLUSTER SETTING jobs.debug.pausepoints = ''`)
+				sqlDB.Exec(t, `RESUME JOB $1`, restoreJobID)
+				jobutils.WaitForJobToSucceed(t, sqlDB, jobspb.JobID(restoreJobID))
+			},
+			checkpointsIntroducedSpans: true,
+		},
+	} {
+		testCase := testCase
+		t.Run(testCase.name, func(t *testing.T) {
+
+			ctx := context.Background()
+
+			// We turn on the following testing knobs during the backup whose
+			// checkpoints we want to track.k
+			exportedSpanCount := make(map[string]int)
+			var logExportSpanCount bool
+			var holdUpCheckpointing bool
+			waitForProgress := make(chan struct{})
+			waitForPause := make(chan struct{})
+
+			params := base.TestClusterArgs{}
+			var mu syncutil.Mutex
+			knobs := base.TestingKnobs{
+				SQLExecutor: &sql.ExecutorTestingKnobs{
+					AfterBackupCheckpoint: func() {
+						mu.Lock()
+						if holdUpCheckpointing && len(exportedSpanCount) > 0 {
+							close(waitForProgress)
+							<-waitForPause
+							holdUpCheckpointing = false
+						}
+						mu.Unlock()
+					},
+				},
+				DistSQL: &execinfra.TestingKnobs{
+					BackupRestoreTestingKnobs: &sql.BackupRestoreTestingKnobs{
+						RunAfterExportingSpanEntry: func(ctx context.Context, response *kvpb.ExportResponse) {
+							mu.Lock()
+							if logExportSpanCount {
+								if response.Files != nil && len(response.Files) > 0 {
+									exportedSpan := response.Files[0].Span.String()
+									exportedSpanCount[exportedSpan] += 1
+								}
+							}
+							mu.Unlock()
+						},
+					},
+				},
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			}
+			params.ServerArgs = base.TestServerArgs{Knobs: knobs}
+
+			tc, sqlDB, dir, cleanupFn := backupRestoreTestSetupWithParams(t, singleNode, 1,
+				InitManualReplication, params)
+			defer cleanupFn()
+
+			sqlDB.Exec(t, `CREATE DATABASE d`)
+			sqlDB.Exec(t, "USE d")
+
+			testCase.preBackupWorkload(ctx, t, sqlDB)
+
+			logExportSpanCount = true
+			holdUpCheckpointing = true
+			sqlDB.Exec(t, "SET CLUSTER SETTING bulkio.backup.checkpoint_interval = '10ms'")
+			var jobID int
+			sqlDB.QueryRow(t, testCase.backupCmd).Scan(&jobID)
+			<-waitForProgress
+			sqlDB.Exec(t, "PAUSE JOB $1", jobID)
+			jobutils.WaitForJobToPause(t, sqlDB, jobspb.JobID(jobID))
+			close(waitForPause)
+
+			backupDetails := jobutils.GetJobPayload(t, sqlDB, jobspb.JobID(jobID)).GetBackup()
+
+			mem, storage, _, _ := getDummyManifestInputs(ctx, t, backupDetails.URI, dir, tc.Servers[0].ExecutorConfig().(sql.ExecutorConfig))
+			desc, _, err := backupinfo.ReadBackupCheckpointManifest(ctx, mem, storage,
+				backupinfo.BackupManifestCheckpointName, nil, nil)
+			require.NoError(t, err)
+
+			// TODO(msbutler): Ideally, I'd assert that the manifest checkpoint
+			// contains at least one file, but it's very hard to control the order we
+			// receive export responses and the timing we flush the checkpoint.
```

Comment:
Would we get export responses in a known order if we set the backup worker count to 1?

Reply:
We would get a known order, but the minWorkerCount is hardcoded to hardcoded 2. I'm not sure why: https://github.com/cockroachdb/cockroach/blob/db3bccef11456826ea231771b147e513a3a44c3b/pkg/ccl/backupccl/backup_processor.go#L121

Reply:
I could add a testing worker count to fix this, but I'm not sure it's worth the effort. When I stress the test, about half the time we actually exercise the right file order that allows us to fully test introduced span checkpointing.